compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -1593,45 +1599,101 @@ private BObjectType getDistinctObjectType(BLangTypeDefinition typeDefinition, BO
         return definedObjType;
     }
 
-    private BType getCyclicDefinedType(BLangTypeDefinition typeDef, SymbolEnv env) {
-        BUnionType unionType = BUnionType.create(null, new LinkedHashSet<>());
-        unionType.isCyclic = true;
-        Name typeDefName = names.fromIdNode(typeDef.name);
-
-        BTypeSymbol typeDefSymbol = Symbols.createTypeSymbol(SymTag.UNION_TYPE, Flags.asMask(typeDef.flagSet),
-                typeDefName, env.enclPkg.symbol.pkgID, unionType, env.scope.owner,
-                typeDef.name.pos, SOURCE);
-
-        typeDef.symbol = typeDefSymbol;
-        unionType.tsymbol = typeDefSymbol;
-
-        // We define the unionType in the main scope here
-        if (PackageID.isLangLibPackageID(this.env.enclPkg.packageID)) {
+    private void defineTypeInMainScope(BTypeSymbol typeDefSymbol, BLangTypeDefinition typeDef, SymbolEnv env) {
+        if (PackageID.isLangLibPackageID(env.enclPkg.packageID)) {
             typeDefSymbol.origin = BUILTIN;
             handleLangLibTypes(typeDef);
         } else {
-            defineSymbol(typeDef.name.pos, typeDefSymbol);
+            defineSymbol(typeDef.name.pos, typeDefSymbol, env);
         }
+    }
+
+    private BType defineSymbolsForCyclicTypeDefinitions(BLangTypeDefinition typeDef, SymbolEnv env) {
+        Name newTypeDefName = names.fromIdNode(typeDef.name);
+        BTypeSymbol typeDefSymbol;
+        BType newTypeNode;
+
+        BSymbol foundSym = symResolver.lookupSymbolInMainSpace(env, names.fromIdNode(typeDef.name));
+        if (foundSym != symTable.notFoundSymbol) {
+            newTypeNode = foundSym.type;
+            newTypeNode.tsymbol = (BTypeSymbol) foundSym;
+            newTypeNode.flags |= foundSym.flags;
```

Comment:
Do we need these flag set and symbol set lines? I think we already set them when we define these in the scope.